docs/DesignDocs/PluggableAllocators.md

```diff
@@ -0,0 +1,142 @@
+Pluggable Allocators
+=====
+
+Open Enclave provides a default memory allocator for enclaves.
+This default allocator, [dlmalloc (Doug Lea Malloc)](http://gee.cs.oswego.edu/dl/html/malloc.html), is not performant in multi-thread applications.
+
+There are various high-performance memory allocators that try to optimize, among other things, multi-thread scenarios:
+- [The GNU Allocator](https://www.gnu.org/software/libc/manual/html_node/The-GNU-Allocator.html) which is based on `pthreads malloc`, which is in turn based on `dlmalloc`.
+- [Thread-Caching Malloc (tcmalloc)](https://gperftools.github.io/gperftools/tcmalloc.html) developed by Google.
+- [snmalloc](https://github.com/Microsoft/snmalloc) developed by Micrsoft Research.
+- and many more. Application specific allocators often outperform general-purpose allocators.
+
+
+This document describes the design of a scheme to allow custom allocators to be used in enclaves.
+
+
+Precedents
+------
+
+- The GNU C Library provides a way to replace its malloc implementation, documented in [Replacing Malloc](https://www.gnu.org/software/libc/manual/html_node/Replacing-malloc.html#Replacing-malloc). </br>
+  To replace the allocator, the following replacement functions must be provided `malloc`, `free`, `calloc`, `realloc`, `aligned_alloc`, `malloc_usable_size`, `memalign`, `posix_memalign`, `palloc`, and `valloc`. </br>
+  Replacement happens by statically linking the allocator library before the GNU C Library.
+
+- The CCF team has implemented compiling the Open Enclave SDK to use `snmalloc` (Pull Request #2649). Currently, users must recompile OE SDK with `USE_SNMALLOC` to leverage this.
```

Comment:
Same comment.  Why does it matter which "team" implemented it?   In a permanent design doc, I don't see why that should matter.